node/network/collator-protocol/src/validator_side.rs

```diff
@@ -481,7 +532,11 @@ async fn handle_our_view_change(
 		.cloned()
 		.collect::<Vec<_>>();
 
+	// Update the set of recently removed chain heads.
+	state.recently_removed_heads.clear();
```

Comment:
Not sure how the view change is handled:
* should it be cleared on when `!removed.is_empty()`?
* should we handle `added` [like bitfield-distribution subsystem](https://github.com/paritytech/polkadot/blob/a5d776a13aa8256c00e3a6f7af59c3b5b2192c60/node/network/bitfield-distribution/src/lib.rs#L454-L472)?

Reply:
we want to keep these for another block to be able not to punish late collation messages about these heads.

So if at `t0` there were leaves `[A, B, C]` and we requested collations on say `B`. Due to network stuff this message was delayed and now we are at `t1` and the active leaves are `[D, E]` and `[A, B, C]` are moved to `recently_removed_heads`. When we process the requested collation on `B` we see that it is late since `B` is already in `recently_removed_heads` but that it is sort of legit message still and we do not punish this peer.

The rest is untouched.

Reply:
Right, my question was like is it guaranteed we won't have spurious `ViewChanged` events that don't alter (or at least remove) anything? E.g.
`t0`: `[A, B, C]`
`t1`: `[D, E]`
`t2`: `[D, E]`
if we process `B` after `t2`, we would punish the peer.

Reply:
> is it guaranteed we won't have spurious ViewChanged

If it is not I am afraid we are going to have code broken in all places.

Reply:
So in the beginning these events were intended to be the "heartbeat that drives all subsystems in one synchronised rythm" or something, not sure if the guide contains this line any more though